tests/lib/openstack_query/test_query_parser.py

```diff
@@ -0,0 +1,415 @@
+        """
+        Setup for tests
+        """
+        super().setUp()
+        self.mock_prop_handler = MagicMock()
+        self.mock_prop_handler.get_prop = MagicMock(wraps=self.mock_get_prop)
+        self.instance = QueryParser(self.mock_prop_handler)
+
+    @staticmethod
+    def mock_get_prop(obj, prop):
+        """simply return obj as prop value to compare"""
+        return obj[prop]
+
+    def test_group_by_prop(self):
+        """
+        Tests that property getter function group_by_prop gets the hidden attribute QueryParser._group_by
+        """
+        self.instance._group_by = "prop1"
+        self.assertEqual(self.instance.group_by_prop, "prop1")
+
+    @patch("openstack_query.query_parser.QueryParser._check_prop_valid")
+    def test_parse_sort_by_one_key(self, mock_check_prop_valid):
+        """
+        Tests parse_sort_by method works expectedly - one sort-by key
+        should call check_prop_valid and add it to sort_by attribute
+        """
+        self.instance.parse_sort_by((MockProperties.PROP_1, False))
+        self.assertEqual(dict(self.instance._sort_by), {MockProperties.PROP_1: False})
+        mock_check_prop_valid.assert_called_once_with(MockProperties.PROP_1)
+
+    @patch("openstack_query.query_parser.QueryParser._check_prop_valid")
+    def test_parse_sort_by_many_keys(self, mock_check_prop_valid):
+        """
+        Tests parse_sort_by method works expectedly - two sort-by keys
+        should call check_prop_valid and add it to sort_by attribute
+        """
+        self.instance.parse_sort_by(
+            (MockProperties.PROP_1, False), (MockProperties.PROP_2, True)
+        )
+        self.assertEqual(
+            dict(self.instance._sort_by),
+            {MockProperties.PROP_1: False, MockProperties.PROP_2: True},
+        )
+        mock_check_prop_valid.assert_has_calls(
+            [call(MockProperties.PROP_1), call(MockProperties.PROP_2)]
+        )
+
+    @parameterized.expand(
+        [("no include missing", False), ("with include missing", True)]
+    )
+    @patch("openstack_query.query_parser.QueryParser._check_prop_valid")
+    def test_parse_group_by_with_group_ranges(
+        self, _, include_missing, mock_check_prop_valid
+    ):
+        """
+        Tests parse_group_by method works expectedly - with valid group ranges defined
+        should create a dictionary with group names mapped to lambda functions
+            - which return True or False if a given resource false into group
+        """
+
+        mock_group_ranges = {"group1": ["val1", "val2", "val3"], "group2": ["val4"]}
+
+        self.instance.parse_group_by(
+            group_by="prop1",
+            group_ranges=mock_group_ranges,
+            include_missing=include_missing,
+        )
+
+        mock_check_prop_valid.assert_called_once_with("prop1")
+        if not include_missing:
+            self.assertEqual(
+                self.instance._group_mappings.keys(), mock_group_ranges.keys()
```

Comment:
We should just add a getter for this, I can't think of a reason why this needs to be a hidden implementation detail

Reply:
a getter for what? The attribute `group_mappings`?
Why do we need a getter? We only use it in `ParseQuery`

Reply:
that makes sense - never considered tests count as consumers of the class :+1: